clang/include/clang/AST/ExprCXX.h

```diff
@@ -4523,11 +4522,11 @@ class SubstNonTypeTemplateParmExpr : public Expr {
   SubstNonTypeTemplateParmExpr(QualType Ty, ExprValueKind ValueKind,
                                SourceLocation Loc, Expr *Replacement,
                                Decl *AssociatedDecl, unsigned Index,
-                               std::optional<unsigned> PackIndex, bool RefParam)
+                               UnsignedOrNone PackIndex, bool RefParam)
       : Expr(SubstNonTypeTemplateParmExprClass, Ty, ValueKind, OK_Ordinary),
         Replacement(Replacement),
         AssociatedDeclAndRef(AssociatedDecl, RefParam), Index(Index),
-        PackIndex(PackIndex ? *PackIndex + 1 : 0) {
+        PackIndex(PackIndex.toInternalRepresentation()) {
```

Comment:
Can you explain this one here?  I'm missing SOMETHING.

Reply:
The pack index here is an unsigned bitfield on the class, and we can't have bitfields of class type unfortunately :)

So we translate the UnsignedOrNone to internal rep in order to store in bitfield, and do the reverse when retrieving it.

Reply:
AH!  Thank you, I missed that.  Does that make my suggestion to use `unsigned-max` as the null-rep value not possible?

Reply:
Yeah it would make things much more error-prone.